File: Swift/1365_how_many_numbers_are_smaller_than_the_current_number.swift
class Solution1 {
    func smallerNumbersThanCurrent(_ nums: [Int]) -> [Int] {
        let n = nums.count
        var answer = [Int]()
        
        // For each number in the array, count how many numbers are smaller than it by comparing it with every other number.
        for i in 0..<n {
            var current = 0
            for j in 0..<n {
                if nums[j] < nums[i] {
                    current += 1
                }
            }
            answer.append(current)
        }
        
        return answer
    }
}

class Solution2 {
    func smallerNumbersThanCurrent(_ nums: [Int]) -> [Int] {
        let temp = nums.sorted()
        var mapping = [Int: Int]()
        
        // Save the index of the first occurrence of each number in the sorted array
        for (i, val) in temp.enumerated() {
            if mapping[val] == nil {
                mapping[val] = i
            }
        }
        
        // Map each number in the original array to its corresponding index in the sorted array
        return nums.map { mapping[$0] ?? 0 }
    }
}

File: Python/1365_how_many_numbers_are_smaller_than_the_current_number.py
# O(n^2) solution
class Solution1:
    def smallerNumbersThanCurrent(self, nums: List[int]) -> List[int]:
        n = len(nums)
        answer = []

        # For each number in the array, count how many numbers are smaller than it by comparing it with every other number.
        for i in range(n):
            current = 0
            for j in range(n):
                if nums[j] < nums[i]:
                    current += 1

            answer.append(current)

        return answer

# O(n log n) solution
class Solution2:
    def smallerNumbersThanCurrent(self, nums: List[int]) -> List[int]:
        temp = sorted(nums)
        mapping = {}
        
        # Save the index of the first occurrence of each number in the sorted array
        for i, val in enumerate(temp):
            if val not in mapping:
                mapping[val] = i
                
        # Map each number in the original array to its corresponding index in the sorted array
        return [mapping[num] for num in nums]